src/tree/hist/histogram.h

```diff
@@ -180,47 +214,100 @@ class HistogramBuilder {
+
+      // Perform AllGather
+      auto hist_vec = std::vector<std::int8_t>(hist_data.data(),
+                                               hist_data.data() + hist_data.size());
+      auto hist_entries = collective::AllgatherV(hist_vec);
+      // Call interface here to post-process the messages
+      std::vector<double> hist_aggr = processor_instance->HandleAggregation(
+              hist_entries.data(), hist_entries.size());
+
+      // Update histogram for label owner
+      if (collective::GetRank() == 0) {
+        // iterator of the beginning of the vector
+        auto it = reinterpret_cast<double *>(this->hist_[first_nidx].data());
+        // iterate through the hist vector of the label owner
+        for (std::size_t i = 0; i < n; i++) {
+          // get the sum of the entries from all ranks
+          double hist_sum = 0.0;
+          for (std::size_t rank_idx = 0; rank_idx < hist_aggr.size()/n; rank_idx++) {
+            int flat_idx = rank_idx * n + i;
+            hist_sum += hist_aggr[flat_idx];
+          }
+          // update rank 0's record with the global histogram
+          *it = hist_sum;
+          it++;
+        }
+      }
     }
 
     common::BlockedSpace2d const &subspace =
-        nodes_to_trick.size() == nodes_to_build.size()
-            ? space
-            : common::BlockedSpace2d{nodes_to_trick.size(),
-                                     [&](std::size_t) { return n_total_bins; }, 1024};
+      nodes_to_trick.size() == nodes_to_build.size()
+      ? space
+      : common::BlockedSpace2d{nodes_to_trick.size(),
+                               [&](std::size_t) { return n_total_bins; }, 1024};
     common::ParallelFor2d(
-        subspace, this->n_threads_, [&](std::size_t nidx_in_set, common::Range1d r) {
-          auto subtraction_nidx = nodes_to_trick[nidx_in_set];
-          auto parent_id = p_tree->Parent(subtraction_nidx);
-          auto sibling_nidx = p_tree->IsLeftChild(subtraction_nidx) ? p_tree->RightChild(parent_id)
+      subspace, this->n_threads_, [&](std::size_t nidx_in_set, common::Range1d r) {
+        auto subtraction_nidx = nodes_to_trick[nidx_in_set];
+        auto parent_id = p_tree->Parent(subtraction_nidx);
+        auto sibling_nidx = p_tree->IsLeftChild(subtraction_nidx) ? p_tree->RightChild(parent_id)
                                                                     : p_tree->LeftChild(parent_id);
-          auto sibling_hist = this->hist_[sibling_nidx];
-          auto parent_hist = this->hist_[parent_id];
-          auto subtract_hist = this->hist_[subtraction_nidx];
-          common::SubtractionHist(subtract_hist, parent_hist, sibling_hist, r.begin(), r.end());
-        });
+        auto sibling_hist = this->hist_[sibling_nidx];
+        auto parent_hist = this->hist_[parent_id];
+        auto subtract_hist = this->hist_[subtraction_nidx];
+        common::SubtractionHist(subtract_hist, parent_hist, sibling_hist, r.begin(), r.end());
+      });
```

Comment:
Please don't make unnecessary changes.

Reply:
just to make the indent more consistent within this file